File: backend/src/main/java/com/olp/backend/entity/Content.java
package com.olp.backend.entity;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import jakarta.persistence.*;

@JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        include = JsonTypeInfo.As.PROPERTY,
        property = "type"
)
@JsonSubTypes({
        @JsonSubTypes.Type(value = Video.class, name = "video"),
        @JsonSubTypes.Type(value = Quiz.class, name = "quiz"),
        @JsonSubTypes.Type(value = Assignment.class, name = "assignment")
})
@Entity
@Inheritance(strategy = InheritanceType.JOINED)
@DiscriminatorColumn(name = "content_type")
public abstract class Content {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    @Column(name = "title")
    private String title;
    @ManyToOne
    @JoinColumn(name = "section_id")
    @JsonIgnoreProperties({"contents"})
    private Section section;

    public Content() {
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public Section getSection() {
        return section;
    }

    public void setSection(Section section) {
        this.section = section;
    }
}

File: frontend/src/models/Content.ts
export interface Content {
    id: number,
    title: string,
    type: 'video' | 'quiz' | 'assignment'
}

export interface Video extends Content {
    type: 'video',
    url: string
}

export interface Quiz extends Content {
    type: 'quiz',
    questions: Question[]
}

export interface Assignment extends Content {
    type: 'assignment',
    description: string
}

export type ContentType = Video | Quiz | Assignment

interface Question {
    id: number,
    text: string,
    options: string[],
    correctOption: string
}